Added a way to sort comments on an article

vishaldeepak opened the pull request:
## What type of PR is this? (check all applicable)

- [*] Feature

## Description

Add a way to sort comments. Comments can be sorted in three ways
- latest
- oldest
- score

Currently deals Point 3 of the issue 
User, post-level

## Related Tickets & Documents
Deals with #16133 

## QA Instructions, Screenshots, Recordings
How it looks :
<img width="912" alt="Screenshot 2022-02-25 at 12 20 02 PM" src="https://user-images.githubusercontent.com/9512872/155750846-a1efa96f-2973-43b9-918d-fe88e72556a8.png">

### UI accessibility concerns?

The view was followed using [this comment](https://github.com/forem/forem/issues/16133#issue-1105555301)
I require svg files for the icons

Want to add accessibility. But would like to confirm the view

Currently
- Selectable via keyboard
- Didnt get any new errors when running Axe tool

## Added/updated tests?

- [ ] No, and this is why: Would like to verify approach first , then add and fix existing cases

**Update**
- Added test cases


## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](gif_link)

vishaldeepak commented on app/controllers/stories_controller.rb:
```diff
@@ -414,4 +414,10 @@ def user_same_as
   end
+
+  def fetch_sort_order
+    return params[:comments_sort] if %w[top latest oldest].include? params[:comments_sort]
```
@citizen428 fixed

nickytonline commented on app/assets/stylesheets/views/comments.scss:
```diff
@@ -275,3 +275,48 @@
+  cursor: pointer;
+}
+
+#comments-sort-dropdown {
```
Refactor (blocking): Styling for a dropdown is not required. Use the CSS class `crayons-dropdown`. Using this class wires up a bunch of functionality to the dropdown to make it more accessible as well.

For margins and padding stick to our CSS variables, e.g. ✅  `padding: var(--su-4);` ❌  `margin-top: 0.25rem;`.

@ludwiczakpawel, do you have any design input here?

ludwiczakpawel replied:
I know im late to the party, but it seems like it has been fixed...?

vishaldeepak replied:
yes, I was waiting on test cases to inform you guys

nickytonline commented on app/views/articles/_full_comment_area.html.erb:
```diff
@@ -2,10 +2,28 @@
       <header class="relative flex justify-between items-center mb-6">
-        <h2 class="crayons-subtitle-1">
-          <%= t("views.articles.comments.subtitle_html",
+        <h2 class="crayons-subtitle-1" id="toggle-comments-sort-dropdown" aria-controls="comments-sort-dropdown" aria-expanded="false" aria-label="Toggle Comment Sort Order dropdown menu" aria-haspopup="true">
```
Refactor (blocking): A header cannot be used as the trigger for the dropdown as it causes accessibility issues. A sort button to the right of the header as the trigger will be more accessible. cc: @aitchiss

aitchiss replied:
Checking out the designs on https://github.com/forem/forem/issues/16133, I think you can use an [icon only button](https://storybook.forem.com/?path=/story/components-buttons--icon-only) with the `expand.svg` icon.

In terms of accessibility, be sure to give the button an `aria-label` that describes the function ("sort comments" would work). We also have a helper to initialize dropdown expand/collapse behaviours in JS (incuding accessibility features) - check out https://github.com/forem/forem/blob/main/app/javascript/utilities/dropdownUtils.js#L118

nickytonline commented on app/views/articles/_full_comment_area.html.erb:
```diff
@@ -2,10 +2,28 @@
         </h2>
+        <div id="comments-sort-dropdown">
+          <h4 class="comments-sort-title">Sort discussion: </h4>
+render "comments/sort_option",
```
Refactor (blocking): There are missing `<%=` for rendering partials.

```suggestion
<%= render "comments/sort_option",
```

![CleanShot 2022-03-07 at 10 27 22](https://user-images.githubusercontent.com/833231/157064219-de415731-8102-4319-927b-8f074e0199f0.png)

vishaldeepak replied:
thats really odd, must be some auto-fix while committing. Will fix this

nickytonline reviewed (changes requested):
@vishaldeepak, thanks so much for the PR. A dropdown is a right approach, but there are some issues with the implementation. I've left some feedback.

vishaldeepak commented:
@nickytonline @ludwiczakpawel  Request changes made , and added relevant test cases added. 
I was unable to start cypress tests locally. Ran into errors. The [contributing docs](https://developers.forem.com/contributing-guide/forem/) don't mention how to run this tests I suppose. I fixed existing test cases for cypress , since that was just a string change, though was unable to add a new tests on them.

aitchiss commented on app/views/comments/_sort_option.html.erb:
```diff
@@ -0,0 +1,14 @@
+<div class="comment-sort-option">
```
> would a `<nav />` with an `aria-label` (to distinguish it from other `<nav />`s) that contains a list of hyperlinks make more sense?

Yes, I think in this case that's most semantic/accessible. Not all lists of links _need_ to be in a `nav`, but I think in this case it could be helpful since they are a very distinct group. So what we want is something like:

```html
<nav aria-labelledby="sort-comments-heading">
  <h3 id="sort-comments-heading">Sort discussion:</h3>
  <ul>
    <li>
      <a href="<link to the sort option>" aria-describedby="<sort option name>-description-text">sort option text</a>
      <span id="<sort option name>-description-text">description text for sort option</span>
    </li>
  <ul>
</nav>
```
This is just pseudo-code, but essentially we want a list of links inside a `nav`, and we want all descriptions to be linked to the link itself with `aria-describedby`. `Top`, `Latest`, etc should not be headings.

Also, `aria-current="page"` should be added to the `<a>` tag when that item is the currently selected sort option (we could even control the appearance of the checkmark icon in CSS with the `[aria-current="page"]` selector)

Let me know if I can provide any more clarification on this!

vishaldeepak replied:
@aitchiss Are you implying that we also load the svg from CSS? 

How would you set color to the svg via css. For example I tried some code 

HTML
```
<li class="comment-sort-option">
  <a href="/admin_mcadmin/this-is-title-2m9f?comments_sort=top" class="comment-sort-option__header first-nav-link" aria-describedby="top-description-text" aria-current="page">
    Top
  </a>
  <span id="top-description-text" class="crayons-field__description">Most upvoted and relevant comments will be first</span>
</li>
```

Css
```
    [aria-current="page"]::before{
      content: '';
      background-image: url(image_path('checkmark.svg'));
      position: relative;
      display: block;
      background-size: 28px 28px;
      height: 28px;
      width: 28px;
    }
```
The problem is that the svg loads but not in the font-color we want.  There's a way I found on [stack](https://stackoverflow.com/a/42317014/3480449).  Should we go that route? 

Other way would be to load the svg in html, but we might end up in code that might be undesirable
```
<li class="comment-sort-option">
  <span class="sort-selected">
      <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" class="crayons-icon" role="img" aria-labelledby="abks7e4kr1ixjsoqqloov97nkecs1rbl"><title id="abks7e4kr1ixjsoqqloov97nkecs1rbl">Selected Sort Option</title>
    <path d="M10 15.172l9.192-9.193 1.415 1.414L10 18l-6.364-6.364 1.414-1.414 4.95 4.95z"></path>
</svg>

  </span>
  <span>
    <a href="/admin_mcadmin/this-is-title-2m9f?comments_sort=top" class="comment-sort-option__header first-nav-link" aria-describedby="top-description-text" aria-current="page">
      Top
    </a>
    <span id="top-description-text" class="crayons-field__description">Most upvoted and relevant comments will be first</span>
  </span>
</li>
```

aitchiss replied:
Yes sorry I wasn't very clear in my suggestion! We can load the svg in the `html.erb` file, using the `crayons_icon_tag` helper function, adding it inside the link, and including a class name like `sort-selected-icon`, keeping it hidden by default (using the helper class name 'hidden'). We could then reveal it using something like

```scss
[aria-current="page"] {
  .sort-selected {
    display: inline-block;
  }
}
```

aitchiss replied:
(but if showing/hiding the icon in this way adds too much complexity, I don't mind going another direction - it was just an idea! 🙂 )

vishaldeepak replied:
changes made.

aitchiss commented on app/assets/stylesheets/views/comments.scss:
```diff
@@ -275,3 +275,59 @@
+}
+
+#comments-sort-dropdown-container {
+  z-index: var(--z-dropdown);
```
(UI bug / blocking) - This applies an elevated z-index to the whole area, including the heading, when we only want it to apply to the actual dropdown. The result is that it's possible to scroll the heading into the top nav bar:

![heading present in top nav bar](https://user-images.githubusercontent.com/20773163/159685190-fb4cc0fa-b309-4cc6-83a9-2714d9bb2a4a.png)

You should be able to rely on the `crayons-dropdown` class on the dropdown content itself to control `z-index`, but let me know if you have any problems with that

vishaldeepak replied:
That was my initial thought as well. But it seemed to cause a UI issue. 
<img width="379" alt="Screenshot 2022-03-26 at 2 00 10 PM" src="https://user-images.githubusercontent.com/9512872/160231475-b9b5a37b-c59e-4948-a509-1399084e4eec.png">

vishaldeepak replied:
@aitchiss do let me know if you figure this one out

aitchiss commented on app/assets/stylesheets/views/comments.scss:
```diff
@@ -275,3 +275,59 @@
+#comments-sort-dropdown-container {
+  z-index: var(--z-dropdown);
+
+  .comments-sort-dropdown {
+    left: 0;
+    padding: var(--su-4);
+    padding-right: var(--su-6);
+    width: 310px;
+
+    &__list {
+      list-style-type: none;
+    }
+  }
```
refactor (blocking) - I think almost all of these you don't need, as they're applied by the `crayons-dropdown` class. The only one I think you require is the padding, and you can add this with a helper classname directly on the HTML element - `p-4`

vishaldeepak replied:
good catch. But without the width property, the end result is a bit different from what was posted in the issue.
<img width="468" alt="Screenshot 2022-03-26 at 2 10 38 PM" src="https://user-images.githubusercontent.com/9512872/160231846-4b045894-8280-4ef4-9459-e77d6d4cb649.png">

aitchiss commented on app/assets/stylesheets/views/comments.scss:
```diff
@@ -275,3 +275,59 @@
+  }
+
+  .comment-sort-option {
+    margin-top: var(--su-2);
+    padding-left: var(--su-7);
```
refactor (not-blocking) - prefer the helper utility classes `mt-2 pl-7` on the HTML element

aitchiss commented on app/assets/stylesheets/views/comments.scss:
```diff
@@ -275,3 +275,59 @@
+    &__header {
+      color: var(--base-90);
+      font-weight: var(--fw-medium);
+      display: block;
```
refactor (blocking) - I don't think this line is needed

aitchiss commented on app/assets/stylesheets/views/comments.scss:
```diff
@@ -275,3 +275,59 @@
+      display: block;
+    }
+
+    .sort-selected {
+      flex-basis: auto;
+      min-width: 40px;
+    }
```
refactor (blocking) - this class doesn't seem to be used anywhere, so let's delete it 
```suggestion
```

aitchiss commented on app/assets/stylesheets/views/comments.scss:
```diff
@@ -275,3 +275,59 @@
+
+    .comment-sort-option__header {
+      .crayons-icon {
+        display: none;
```
refactor (blocking) - I think you went a different route with this in the end, and the icon is only added in the HTML next to the currently selected option. So we no longer need this CSS property, or the related `aria-current` selector

```suggestion
```

aitchiss commented on app/assets/stylesheets/views/comments.scss:
```diff
@@ -275,3 +275,59 @@
+
+      &[aria-current='page'] {
+        font-weight: var(--fw-bold);
+
+        .crayons-icon {
+          display: block;
+        }
```
```suggestion
```

(as above)

aitchiss commented on app/assets/stylesheets/views/comments.scss:
```diff
@@ -275,3 +275,59 @@
+    }
+  }
+
+  .comments-sort-title {
+    margin-bottom: var(--su-3);
+  }
```
refactor (blocking) - as there's only one property here, let's favour using helper utility classes on the element itself. You should be able to use `mb-3` classname on the element to get the same effect

aitchiss reviewed (changes requested):
Thanks for your work on this one! It's working really well, but I've requested a few refactoring/cleanup changes that will help make this more robust and maintainable in future.

I've also asked that we add some Cypress tests for this flow, but if you need assistance with that, just let me know and either myself or someone else from the team can chip in with those!

vishaldeepak commented on spec/system/articles/user_visits_an_article_spec.rb:
```diff
@@ -184,4 +184,39 @@
     end
   end
+
+  describe "sort comments" do
```
Hey I need  help to setup cypress on local. I was unable to start cypress tests locally. Ran into errors. The [contributing docs](https://developers.forem.com/contributing-guide/forem/) don't mention how to run this tests I suppose. 
I tried starting the server locally via `rake cypress:open`
But I run into an error
<img width="670" alt="Screenshot 2022-03-26 at 11 25 19 PM" src="https://user-images.githubusercontent.com/9512872/160251557-c0ba914e-d9b7-4766-9da1-2184100f84fe.png">

aitchiss replied:
Ah can you try running the specs using the command `yarn e2e`. You'll see a message in the console asking you if you want to set up your database before running, and please type 'Y' to confirm that yep we want this! Hopefully that should resolve your issues.

[The E2E testing docs for Forem are here](https://developers.forem.com/tests/e2e-tests) - let me know if you notice things missing from them though, as we're always working on improving these

vishaldeepak replied:
added tests in cypress, let me know If should remove the the above test?

aitchiss replied:
Yeah I think we can remove this one now, as it's well covered in Cypress 🔥

aitchiss commented on app/javascript/packs/commentDropdowns.js:
```diff
@@ -1,4 +1,5 @@
 import { addSnackbarItem } from '../Snackbar';
+import { initializeMemberMenu } from '../topNavigation/utilities';
```
This file is related to the top member navigaton, and isn't something we need to use here - trying to repurpose this is part of what is causing the stacking issues

Instead, see below, and comments in `app/views/articles/_full_comment_area.html.erb` for how we can initialize the dropdown

aitchiss commented on app/javascript/packs/commentDropdowns.js:
```diff
@@ -69,6 +74,18 @@ const initializeArticlePageDropdowns = () => {
     }
   }
+
+  const dropdownContent = document.getElementById(
+    'comments-sort-dropdown-container',
+  );
+  const dropdownElement = document.getElementById(
+    'toggle-comments-sort-dropdown',
+  );
+
+  if (dropdownContent && !dropdownContent.dataset.initialized) {
+    initializeMemberMenu(dropdownContent, dropdownElement);
+    dropdownContent.dataset.initialized = true;
+  }
```
```suggestion
  const sortCommentsDropdownTrigger = document.getElementById(
    'toggle-comments-sort-dropdown',
  );

  if (!sortCommentsDropdownTrigger.dataset.initialized) {
     initializeDropdown({
       triggerElementId: 'toggle-comments-sort-dropdown',
       dropdownContentId: 'comments-sort-dropdown-container',
     });
    sortCommentsDropdownTrigger.dataset.initialized = true
  }
```

This refactors to use the `initializeDropdown` helper used elsewhere. This helper is designed to work with `crayons-dropdown` classes 🙂

vishaldeepak replied:
thanks so much, I made another change that was a bit easier

aitchiss commented on app/views/articles/_full_comment_area.html.erb:
```diff
@@ -2,10 +2,40 @@
-          <%= t("views.articles.comments.subtitle_html",
-                num: tag.span(t("views.articles.comments.num", num: @article.comments_count), class: "js-comments-count", data: { comments_count: @article.comments_count })) %>
-        </h2>
+        <div class="crayons-header__menu" id="comments-sort-dropdown-container">
+          <div class="crayons-subtitle-1" style="display: flex">
+            <h4>
+              <%= t("views.articles.comments.subtitle.#{@comments_order}_html",
+                    num: tag.span(t("views.articles.comments.num", num: @article.comments_count), class: "js-comments-count", data: { comments_count: @article.comments_count })) %>
+            </h4>
+            <button class="crayons-btn crayons-btn--ghost crayons-btn--icon-left" id="toggle-comments-sort-dropdown" aria-controls="comments-sort-dropdown-container" aria-expanded="false" aria-label="Sort comments" aria-haspopup="true">
+              <%= inline_svg_tag("expand.svg", aria_hidden: true, title: t("views.moderations.actions.admin.icon")) %>
+            </button>
+          </div>
+          <nav class="comments-sort-dropdown crayons-dropdown crayons-header__menu__dropdown inline-block" aria-labelledby="sort-comments-heading">
+              <h4 class="comments-sort-title">Sort discussion: </h4>
+              <ul class="comments-sort-dropdown__list">
+                <%= render "comments/sort_option",
+                           sort_order: "top",
+                           sort_title: "Top",
+                           sort_description: "Most upvoted and relevant comments will be first",
+                           show_selected: @comments_order == "top",
+                           html_classes: "first-nav-link" %>
+                <%= render "comments/sort_option",
+                           sort_order: "latest",
+                           sort_title: "Latest",
+                           sort_description: "Most recent comments will be first",
+                           show_selected: @comments_order == "latest",
+                           html_classes: "second-last-nav-link" %>
+                <%= render "comments/sort_option",
+                           sort_order: "oldest",
+                           sort_title: "Oldest",
+                           sort_description: "The oldest comments will be first",
+                           show_selected: @comments_order == "oldest",
+                           html_classes: "last-nav-link" %>
+              </ul>
+          </nav>
+        </div>
```
```suggestion
<div class="flex">
          <h2>
            <%= t("views.articles.comments.subtitle.#{@comments_order}_html",
                num: tag.span(t("views.articles.comments.num", num: @article.comments_count), class: "js-comments-count", data: { comments_count: @article.comments_count })) %>
          </h2>
          <button id="toggle-comments-sort-dropdown" class="crayons-btn crayons-btn--ghost crayons-btn--icon-left" id="toggle-comments-sort-dropdown" aria-controls="comments-sort-dropdown-container" aria-expanded="false" aria-label="Sort comments" aria-haspopup="true">
            <%= inline_svg_tag("expand.svg", aria_hidden: true, title: t("views.moderations.actions.admin.icon")) %>
          </button>
        </div>
        <nav id="comments-sort-dropdown-container" class="crayons-dropdown p-4" aria-labelledby="sort-comments-heading">
          <h4 class="comments-sort-title">Sort discussion: </h4>
          <ul class="comments-sort-dropdown__list">
            <%= render "comments/sort_option",
                        sort_order: "top",
                        sort_title: "Top",
                        sort_description: "Most upvoted and relevant comments will be first",
                        show_selected: @comments_order == "top",
                        html_classes: "first-nav-link" %>
            <%= render "comments/sort_option",
                        sort_order: "latest",
                        sort_title: "Latest",
                        sort_description: "Most recent comments will be first",
                        show_selected: @comments_order == "latest",
                        html_classes: "second-last-nav-link" %>
            <%= render "comments/sort_option",
                        sort_order: "oldest",
                        sort_title: "Oldest",
                        sort_description: "The oldest comments will be first",
                        show_selected: @comments_order == "oldest",
                        html_classes: "last-nav-link" %>
          </ul>
        </nav>
```

Here I have again refactored to remove any conflation with the member top navigation menu, as mentioned in the `commentDropdowns` file. Removing the conflation allows the dropdown to be opened/closed, relying on the `crayons-dropdown` class and without any strange stacking issues (it is expected that the open dropdown would appear over the sticky header when scrolled):

https://user-images.githubusercontent.com/20773163/160593995-3315e4bb-82b6-44e7-9ce9-5b630ecdef6b.mp4

aitchiss reviewed (commented):
Added some comments on how tI managed to resolve the stacking issues locally. It seems the issue is trying to apply the top member nav menu classes/code to this dropdown, which as I mentioned before, isn't what we want to do 🙂 

Let me know if you have any more difficulty!

vishaldeepak commented:
@aitchiss  Thanks so much for your help. I have added cypress test cases and cleand up code. Should I remove `system/articles` spec

aitchiss commented on app/javascript/packs/commentDropdowns.js:
```diff
@@ -21,7 +21,7 @@ const handleCopyPermalink = (closeDropdown) => {
   // Gather all dropdown triggers for comment options and profile previews
   const dropdownTriggers = document.querySelectorAll(
-    'button[id^=comment-dropdown-trigger], button[id^=comment-profile-preview-trigger-]',
+    'button[id^=comment-dropdown-trigger], button[id^=comment-profile-preview-trigger-],#toggle-comments-sort-dropdown',
```
perfect! should we maybe keep the new query selector in line with the existing ones, e.g.
```suggestion
    'button[id^=comment-dropdown-trigger], button[id^=comment-profile-preview-trigger-], button[id^=toggle-comments-sort-dropdown],
```

(this isn't in any way blocking, just a thought 😄 )

aitchiss commented on cypress/integration/seededFlows/articleFlows/sortCommentsOnArticle.spec.js:
```diff
@@ -0,0 +1,88 @@
+
+    // Helper function for pipe command
+    const click = ($el) => $el.click();
+    cy.findByRole('button', { name: 'Sort comments' }).pipe(click).click();
```
(same as above)

aitchiss reviewed (commented):
Nice work adding the Cypress tests - they look great! I'll take this for another little spin now 😄

aitchiss commented:
I noticed a bit of a strange bug, where once I select an item from the dropdown, the page refreshes with the right content, but then next time I click the dropdown button, nothing happens:

https://user-images.githubusercontent.com/20773163/161053408-22e8d1b8-c4a6-4170-8180-263b85a7c6c1.mp4


I had a little investigate and it seems to be this code in `dropdownUtils` causing the issue:

https://github.com/forem/forem/blob/0403e78f08b2de9e90160be721896ab9879c08b9/app/javascript/utilities/dropdownUtils.js#L167-L184

It seems like what it thinks is the `triggerButton` is still the button on the _last_ page, rather than the current one 🙈  I think this is less a bug with your code, and more a bug with the utility function. How would you feel about adding the fix in this PR though? I think all we need to do is to update `clickOutsideListener` to get a fresh handle of the button each time, which would be a bit more robust e.g. `const triggerButton = document.getElementById(triggerElementId);`

vishaldeepak commented:
> I noticed a bit of a strange bug, where once I select an item from the dropdown, the page refreshes with the right content, but then next time I click the dropdown button, nothing happens:
> CleanShot.2022-03-31.at.13.20.49.mp4
> 
> I had a little investigate and it seems to be this code in `dropdownUtils` causing the issue:
> 
> https://github.com/forem/forem/blob/0403e78f08b2de9e90160be721896ab9879c08b9/app/javascript/utilities/dropdownUtils.js#L167-L184
> 
> It seems like what it thinks is the `triggerButton` is still the button on the _last_ page, rather than the current one 🙈 I think this is less a bug with your code, and more a bug with the utility function. How would you feel about adding the fix in this PR though? I think all we need to do is to update `clickOutsideListener` to get a fresh handle of the button each time, which would be a bit more robust e.g. `const triggerButton = document.getElementById(triggerElementId);`

vishaldeepak commented:
@aitchiss 
hey, made changes accordingly. Changing the `clickOustideListener` function, caused a Jest test to fail. `CommentSubscription.test.jsx:62:12`. I had to add another conditional in `clickOustideListener` to make this work.

aitchiss reviewed (approved):
Great work @vishaldeepak! Thank you for making those changes - it's working great now 🔥 

We'll get one more review on this and then we should be good to go 🚀

djuber commented on app/controllers/stories_controller.rb:
```diff
@@ -414,4 +414,10 @@ def user_same_as
   end
+
+  def fetch_sort_order
+    return params[:comments_sort] if Comment::VALID_SORT_OPTIONS.include? params[:comments_sort]
```
refactor, non-blocking: the logic here about which sort options are valid, and the "top" default sort option if not provided, might be moved to a method in the Comment model, passing the `:comment_sort` param to the method and receiving a string back. That keeps the http request details (params provided) in the controller, but the rules about sorting behavior in the model (where you added the constant).

The code smell here for me is reaching into the value of the class's constant, outside of a controller tied to that model like CommentsController.

Something like this might make sense. 

```ruby
class Comment
  def self.sort_order(option)
     # code from fetch_sort_order here    
  end
end
```

I'll merge this as-is, if this logic needs to change in the future we can revisit moving the decision making out of the controller.
